src/main/java/org/hydev/mcpm/client/database/fetcher/LocalDatabaseFetcher.java

```
@@ -0,0 +1,192 @@
+
+    /**
+     * Creates a database fetcher that syncs with a mirror.
+     * The cache directory will default to .mcpm
+     *
+     * @param host The host URL MCPM server mirror that will be used to request database files.
+     *             Example: "http://mcpm.hydev.com"
+     */
+    public LocalDatabaseFetcher(URI host) {
+        this(host, Path.of(".mcpm/"));
+    }
+
+    /**
+     * Creates a database fetcher that syncs with a mirror.
+     *
+     * @param host The host URL MCPM server mirror that will be used to request database files.
+     *             Example: "http://mcpm.hydev.com"
+     * @param cacheDirectory The directory that the cached database and hash file will be stored (as child files).
+     */
+    public LocalDatabaseFetcher(URI host, Path cacheDirectory) {
+        this.host = host;
+        this.cacheDirectory = cacheDirectory;
+    }
+
+    private Request requestTo(String path) {
+        return Request
+            .get(URI.create(host.toString() + "/" + path))
+            .addHeader("Host", host.getHost())
+            .addHeader("User-Agent", USER_AGENT)
+            .addHeader("Accepts", "application/json");
+    }
+
+    @Nullable
+    private String fetchHostHash() {
+        try {
+            var response = requestTo(HASH_FILE_NAME).execute();
+
+            return response.returnContent().asString().trim();
+        } catch (IOException e) {
+            return null;
+        }
+    }
+
+    @Nullable
+    private String fetchLocalHash() {
+        try {
+            return Files.readString(cacheDirectory.resolve(DATABASE_FILE_NAME));
+        } catch (IOException e) {
+            return null;
+        }
+    }
+
+    private boolean localDatabaseIsUpToDate() {
+        var hostHash = fetchHostHash();
+        var localHash = fetchLocalHash();
+
+        // We can't reach the server right now, don't update.
+        if (hostHash == null)
+            return true;
```

Maybe throw a custom exception here so that client can tell the user what's going on

---

Actually, I'd like to keep this as a silent failure. This `return true` doesn't abort anything, it just uses the on-disk database if it's there. I think that would enable users to search/list/check for updates on packages when our server is down or our internet isn't quite working.

---

Oh yea right